src/main/java/at/ac/tuwien/kr/alpha/solver/DefaultSolver.java

```diff
@@ -124,108 +127,124 @@ private BranchingHeuristic chainFallbackHeuristic(Grounder grounder, WritableAss
+			if (searchState.isSearchSpaceCompletelyExplored) {
+				LOGGER.debug("Search space has been fully explored, there are no more answer-sets.");
+				logStats();
+				return false;
 			}
+			ConflictCause conflictCause = propagate();
 			if (conflictCause != null) {
-				// Learn from conflict.
-				LOGGER.debug("Violating assignment is: {}", assignment);
-				Antecedent conflictAntecedent = conflictCause.getAntecedent();
-				NoGood violatedNoGood = new NoGood(conflictAntecedent.getReasonLiterals().clone());
-				// TODO: The violatedNoGood should not be necessary here, but this requires major type changes in heuristics.
-				branchingHeuristic.violatedNoGood(violatedNoGood);
-				if (!afterAllAtomsAssigned) {
-					if (!learnBackjumpAddFromConflict(conflictCause)) {
-						logStats();
-						return false;
-					}
-				} else {
-					LOGGER.debug("Assignment is violated after all unassigned atoms have been assigned false.");
-					conflictsAfterClosing++;
-					if (!treatConflictAfterClosing(conflictAntecedent)) {
-						return false;
-					}
-					afterAllAtomsAssigned = false;
-				}
-			} else if (didChange) {
-				// Ask the grounder for new NoGoods, then propagate (again).
-				LOGGER.trace("Doing propagation step.");
-
+				LOGGER.debug("Conflict encountered, analyzing conflict.");
+				learnFromConflict(conflictCause);
+			} else if (assignment.didChange()) {
+				LOGGER.debug("Updating grounder with new assignments and (potentially) obtaining new NoGoods.");
 				grounder.updateAssignment(assignment.getNewPositiveAssignmentsIterator());
-
-				Map<Integer, NoGood> obtained = grounder.getNoGoods(assignment);
-				didChange = !obtained.isEmpty();
-				if (!ingest(obtained)) {
-					logStats();
-					return false;
-				}
+				getNoGoodsFromGrounderAndIngest();
 			} else if (choose()) {
 				LOGGER.debug("Did choice.");
-				didChange = true;
 			} else if (close()) {
 				LOGGER.debug("Closed unassigned known atoms (assigning FALSE).");
-				afterAllAtomsAssigned = true;
 			} else if (assignment.getMBTCount() == 0) {
-				// NOTE: If we would do optimization, we would now have a guaranteed upper bound.
-				AnswerSet as = translate(assignment.getTrueAssignments());
-				LOGGER.debug("Answer-Set found: {}", as);
-				action.accept(as);
-				logStats();
+				provideAnswerSet(action);
 				return true;
 			} else {
-				LOGGER.debug("Backtracking from wrong choices ({} MBTs).", assignment.getMBTCount());
-				if (!justifyMbtAndBacktrack()) {
-					return false;
-				}
-				afterAllAtomsAssigned = false;
+				backtrackFromMBTsRemaining();
+			}
+		}
+	}
+
+	private void setupLoop() {
+		if (!searchState.hasBeenInitialized) {
+			// Initially, get NoGoods from grounder.
+			performanceLog.initialize();
+			getNoGoodsFromGrounderAndIngest();
+			searchState.hasBeenInitialized = true;
+			return;
+		}
+		// We already found one Answer-Set and are requested to find another one.
+		searchState.afterAllAtomsAssigned = false;
```

Comment:
If I read the code correctly, we reach this part of the code on the second (and subsequent) call(s) to tryAdvance (as in the first call, we'll return after initializing search space). This is not really intuitive from the comment here, I think readability would benefit from a somewhat more detailed explanation on _why_ we only get there after finding an answer set in the comment.

Reply:
Yes, the code is doing that. I split up the method into two separate ones.